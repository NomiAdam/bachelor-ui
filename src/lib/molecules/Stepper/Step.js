import React from 'react';
import PropTypes from 'prop-types';
import Paragraph from '../../atoms/Paragraph/index';
import styled from 'styled-components';
import {lightTheme} from '../../constants/theme';

const StyledStep = styled.li`
  list-style-type: none; 
  float: left; 
  width: ${({stepCount}) => `${ 100 / stepCount}%`}; 
  position: relative; 
  text-align: center;
  font-weight: 600;
  color: ${({isActive}) => isActive ? lightTheme.lightGreen : 'initial' };
  &:before {
    content:counter(step); 
    counter-increment: step; 
    height:30px;
    width:30px; 
    line-height: 30px; 
    border: 2px solid; 
    display: block; 
    text-align: center;
    margin: 0 auto 10px auto; 
    border-radius: 50%; 
    background-color: ${lightTheme.white}; 
    border-color: ${({isActive}) => isActive ? lightTheme.lightGreen : lightTheme.lightGrey };
  }
  &:after {
    content: '';
    position: absolute; 
    width: 100%;
    height: 2px;
    top: 15px; 
    left: -50%; 
    z-index: -1; 
    background-color: ${({isActive}) => isActive ? lightTheme.lightGreen : lightTheme.lightGrey };
 }
 &:first-child:after {
    content: none;
  } 
`;

const Step = ({label, isActive, stepCount, handleClick}) => (
    <StyledStep isActive={isActive} stepCount={stepCount} onClick={handleClick}>
        <Paragraph>
            {label}
        </Paragraph>
    </StyledStep>
);

Step.propTypes = {
    /**
     * Number representing how many step we have
     */
    stepCount: PropTypes.number,
    /**
     * Whether this step is active
     */
    isActive: PropTypes.bool,
    /**
     * String representation of step label
     */
    label: PropTypes.string,
    /**
     * Function handle to handle onClick event
     */
    handleClick: PropTypes.func,
};

export default Step;
