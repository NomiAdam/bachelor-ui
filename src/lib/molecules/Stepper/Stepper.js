import React, {PureComponent} from 'react';
import Container from '../../atoms/Container/index';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import Step from './Step';
import { gt, length } from 'ramda';
import { mapIndexed } from 'ramda-extension';

const StyledStepContainer = styled.ul`
  counter-reset: step; 
  padding: 0;
  margin: 0;
  min-width: 350px;
`;

const renderStep = (handleClick, stepCount, step) => (label, index) => (
    <Step
        key={label}
        // +1 because we are indexing from zero
        handleClick={handleClick(index + 1)}
        stepCount={stepCount}
        label={label}
        isActive={gt(step, index)}
    />
);

const renderSteppersStep = (handleClick, stepCount, step) => mapIndexed(renderStep(handleClick, stepCount, step));


class Stepper extends PureComponent {
    handleStepClick = (step) => () => {
        const { handleClick } = this.props;
        handleClick(step);
    };

    render() {
        const { items, step } = this.props;
        const stepCount = length(items);
        return (
            <Container>
                <StyledStepContainer>
                    {renderSteppersStep(this.handleStepClick, stepCount, step)(items)}
                </StyledStepContainer>
            </Container>
        );
    }
}

Stepper.propTypes = {
    /**
     * Array of all stepping items, note that it should be ordered array
     */
    items: PropTypes.array,
    /**
     * Number representing current step
     */
    step: PropTypes.number.isRequired,
    /**
     * Function handler to handle click on step
     */
    handleClick: PropTypes.func,
};

Stepper.defaultProps = {
    items: [],
};

export default Stepper;
