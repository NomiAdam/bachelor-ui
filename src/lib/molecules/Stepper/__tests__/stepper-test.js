import React from 'react';
import Stepper from '../Stepper';
import { shallow } from 'enzyme';

describe('Stepper', () => {
    it('should be defined', () => {
        expect(Stepper).toBeDefined();
    });
    it('should match the snapshot', () => {
        expect(shallow(<Stepper />)).toMatchSnapshot();
    });
});
