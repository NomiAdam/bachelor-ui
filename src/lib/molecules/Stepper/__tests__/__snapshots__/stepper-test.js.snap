// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stepper should match the snapshot 1`] = `
<styled.div
  alignItems="center"
  flexDirection="column"
  height="100%"
  justifyContent="center"
  width="100%"
>
  <styled.ul />
</styled.div>
`;
